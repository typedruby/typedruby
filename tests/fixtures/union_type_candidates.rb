def get_length[T]((String | Array::[T]) x) => Integer; end

def test1 => nil
  get_length("foo")
  get_length([1, 2, 3])
  nil
end

class P::[T]
  def initialize(T x) => nil; end
end

def wrap[T]((T | P::[T]) p) => P::[T]; end

def test2 => nil
  reveal_type(wrap(123))
  reveal_type(wrap(P.new(123)))
  nil
end

def test3 => nil
  reveal_type(wrap((nil : :any)))
  nil
end

def wrap2[T]((Array::[T] | P::[T]) x) => P::[T]
end

def test4 => nil
  reveal_type(wrap2([123]))
  reveal_type(wrap2(P.new("foo")))
  reveal_type(wrap2((nil : :any)))
  nil
end

def ambiguous_var[T, U]((T | U) x) => nil; end

def test5 => nil
  ambiguous_var(123)
  nil
end

module A
end

module B
end

class C
  include A
  include B
end

def ambiguous_module((A | B) x) => nil; end

def test6 => nil
  ambiguous_module(C.new)
end
