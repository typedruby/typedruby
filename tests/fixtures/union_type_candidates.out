error: Revealed type is: P::[Integer]

        @ __ROOT__/tests/fixtures/union_type_candidates.rb:16
     16 |    reveal_type(wrap(123))
                         ^^^^^^^^^ expression

error: Revealed type is: P::[Integer]

        @ __ROOT__/tests/fixtures/union_type_candidates.rb:17
     17 |    reveal_type(wrap(P.new(123)))
                         ^^^^^^^^^^^^^^^^ expression

error: Revealed type is: P::[t2]

        @ __ROOT__/tests/fixtures/union_type_candidates.rb:22
     22 |    reveal_type(wrap((nil : :any)))
                         ^^^^^^^^^^^^^^^^^^ expression

error: Revealed type is: P::[Integer]

        @ __ROOT__/tests/fixtures/union_type_candidates.rb:30
     30 |    reveal_type(wrap2([123]))
                         ^^^^^^^^^^^^ expression

error: Revealed type is: P::[String]

        @ __ROOT__/tests/fixtures/union_type_candidates.rb:31
     31 |    reveal_type(wrap2(P.new("foo")))
                         ^^^^^^^^^^^^^^^^^^^ expression

error: Revealed type is: P::[t23]

        @ __ROOT__/tests/fixtures/union_type_candidates.rb:32
     32 |    reveal_type(wrap2((nil : :any)))
                         ^^^^^^^^^^^^^^^^^^^ expression
